src/main/java/com/rapid7/client/dcerpc/io/PacketOutput.java

```diff
@@ -100,29 +110,48 @@ public void writeString(final String string, final boolean nullTerminate)
 
         if (string != null) {
             writeReferentID();
-            writeInt(maximumChars);
-            writeInt(0);
-            writeInt(currentChars);
-            writeChars(string);
-
-            if (nullTerminate) {
-                writeShort((short) 0);
-            }
-
-            align();
+            writeString(string, maximumChars, currentChars, nullTerminate);
         } else {
             writeNull();
         }
     }
 
     public void writeStringRef(final String string, final boolean nullTerminate)
         throws IOException {
-        if (string != null) {
-            writeReferentID();
-            writeString(string, nullTerminate);
+        writeReferentID();
+        writeString(string, nullTerminate);
+    }
+
+    public void writeString(final String string, final boolean nullTerminate)
+        throws IOException {
+        final int maximumChars;
+        final int currentChars;
+
+        if (string == null) {
+            maximumChars = 1;
+            currentChars = 1;
         } else {
-            writeNull();
+            maximumChars = string.length() + (nullTerminate ? 1 : 0);
+            currentChars = string.length() + (nullTerminate ? 1 : 0);
         }
+
+        writeString(string, maximumChars, currentChars, nullTerminate);
+    }
+
+    private void writeString(final String string, final int maximumChars, final int currentChars, final boolean nullTerminate)
```

Comment:
This method is relying to much on it's caller.

Reply:
This method is private, only useful for making some of the above writeString* functions simpler.